File: hydra-haskell/src/main/haskell/Hydra/Impl/Haskell/Dsl/Literals.hs
module Hydra.Impl.Haskell.Dsl.Literals where

import Hydra.All

import Data.Int


bigfloat :: Double -> Literal
bigfloat = float . FloatValueBigfloat

bigint :: Integer -> Literal
bigint = integer . IntegerValueBigint . fromIntegral

binary :: String -> Literal
binary = LiteralBinary

boolean :: Bool -> Literal
boolean = LiteralBoolean

expectBinary :: Literal -> Flow s String
expectBinary v = case v of
  LiteralBinary b -> pure b
  _ -> unexpected "binary" v

expectBoolean :: Literal -> Flow s Bool
expectBoolean v = case v of
  LiteralBoolean b -> pure b
  _ -> unexpected "boolean" v

expectFloat32 :: Literal -> Flow s Float
expectFloat32 v = case v of
  LiteralFloat (FloatValueFloat32 f) -> pure f
  _ -> unexpected "float32" v

expectFloat64 :: Literal -> Flow s Double
expectFloat64 v = case v of
  LiteralFloat (FloatValueFloat64 f) -> pure f
  _ -> unexpected "float64" v

expectInt32 :: Literal -> Flow s Int
expectInt32 v = case v of
  LiteralInteger (IntegerValueInt32 i) -> pure i
  _ -> unexpected "int32" v

expectInt64 :: Literal -> Flow s Integer
expectInt64 v = case v of
  LiteralInteger (IntegerValueInt64 i) -> pure i
  _ -> unexpected "int64" v

expectString :: Literal -> Flow s String
expectString v = case v of
  LiteralString s -> pure s
  _ -> unexpected "string" v

float32 :: Float -> Literal
float32 = float . FloatValueFloat32

float64 :: Double -> Literal
float64 = float . FloatValueFloat64

float :: FloatValue -> Literal
float = LiteralFloat

int16 :: Int16 -> Literal
int16 = integer . IntegerValueInt16 . fromIntegral

int32 :: Int -> Literal
int32 = integer . IntegerValueInt32

int64 :: Int64 -> Literal
int64 = integer . IntegerValueInt64 . fromIntegral

int8 :: Int8 -> Literal
int8 = integer . IntegerValueInt8 . fromIntegral

integer :: IntegerValue -> Literal
integer = LiteralInteger

string :: String -> Literal
string = LiteralString

uint16 :: Integer -> Literal
uint16 = integer . IntegerValueUint16 . fromIntegral

uint32 :: Integer -> Literal
uint32 = integer . IntegerValueUint32 . fromIntegral

uint64 :: Integer -> Literal
uint64 = integer . IntegerValueUint64 . fromIntegral

uint8 :: Integer -> Literal
uint8 = integer . IntegerValueUint8 . fromIntegral

File: heads/python/src/main/python/hydra/dsl/literals.py
"""A DSL for constructing Hydra literal values in Python."""

from hydra.core import (
    Literal,
    FloatValue,
    IntegerValue,
    LiteralBinary,
    LiteralBoolean,
    LiteralFloat,
    LiteralInteger,
    LiteralString,
    FloatValueBigfloat,
    IntegerValueBigint,
    FloatValueFloat32,
    FloatValueFloat64,
    IntegerValueInt16,
    IntegerValueInt32,
    IntegerValueInt64,
    IntegerValueInt8,
    IntegerValueUint16,
    IntegerValueUint32,
    IntegerValueUint64,
    IntegerValueUint8,
)

from decimal import Decimal


def binary(value: bytes) -> Literal:
    """Create a binary data literal.

    Example: binary(b'\\x48\\x65\\x6C\\x6C\\x6F')
    """
    return LiteralBinary(value)


def boolean(value: bool) -> Literal:
    """Create a boolean literal.

    Example: boolean(True)
    """
    return LiteralBoolean(value)


def string(value: str) -> Literal:
    """Create a string literal.

    Example: string("hello world")
    """
    return LiteralString(value)


def float_(value: FloatValue) -> Literal:
    """Create a floating-point literal with specified precision.

    Example: float_(FloatValueFloat32(3.14))
    """
    return LiteralFloat(value)


def float32(value: float) -> Literal:
    """Create a 32-bit floating point literal.

    Example: float32(3.14)
    """
    return float_(FloatValueFloat32(value))


def float64(value: float) -> Literal:
    """Create a 64-bit floating point literal.

    Example: float64(3.14159265359)
    """
    return float_(FloatValueFloat64(value))


def bigfloat(value: float) -> Literal:
    """Create an arbitrary-precision floating point literal.

    Example: bigfloat(3.14159265359)
    """
    # Use str(value) to avoid full binary precision conversion
    return float_(FloatValueBigfloat(Decimal(str(value))))


def integer(value: IntegerValue) -> Literal:
    """Create an integer literal with specified bit width.

    Example: integer(IntegerValueInt32(42))
    """
    return LiteralInteger(value)


def int8(value: int) -> Literal:
    """Create an 8-bit signed integer literal.

    Example: int8(127)
    """
    return integer(IntegerValueInt8(value))


def int16(value: int) -> Literal:
    """Create a 16-bit signed integer literal.

    Example: int16(32767)
    """
    return integer(IntegerValueInt16(value))


def int32(value: int) -> Literal:
    """Create a 32-bit signed integer literal.

    Example: int32(42)
    """
    return integer(IntegerValueInt32(value))


def int64(value: int) -> Literal:
    """Create a 64-bit signed integer literal.

    Example: int64(9223372036854775807)
    """
    return integer(IntegerValueInt64(value))


def bigint(value: int) -> Literal:
    """Create an arbitrary-precision integer literal.

    Example: bigint(9223372036854775808)
    """
    return integer(IntegerValueBigint(value))


def uint8(value: int) -> Literal:
    """Create an 8-bit unsigned integer literal.

    Example: uint8(255)
    """
    return integer(IntegerValueUint8(value))


def uint16(value: int) -> Literal:
    """Create a 16-bit unsigned integer literal.

    Example: uint16(65535)
    """
    return integer(IntegerValueUint16(value))


def uint32(value: int) -> Literal:
    """Create a 32-bit unsigned integer literal.

    Example: uint32(4294967295)
    """
    return integer(IntegerValueUint32(value))


def uint64(value: int) -> Literal:
    """Create a 64-bit unsigned integer literal.

    Example: uint64(18446744073709551615)
    """
    return integer(IntegerValueUint64(value))
